feat: Adding Nvidia model training components.

maheshrajamani opened the pull request:
Adding model training components based on Nvidia NeMo micro service.

maheshrajamani committed:
Adding Nvidia model training components.

autofix-ci[bot] committed:
[autofix.ci] apply automated fixes

maheshrajamani committed:
Merge branch 'main' into model-training

maheshrajamani committed:
Fix for issue reported by the code style check

maheshrajamani committed:
Fixed merge conflict

autofix-ci[bot] committed:
[autofix.ci] apply automated fixes

maheshrajamani committed:
Fixed the error returned for evaluator code styling

maheshrajamani committed:
Merge branch 'model-training' of https://github.com/maheshrajamani/langflow into model-training

maheshrajamani committed:
Fixed the error returned for evaluator code styling

autofix-ci[bot] committed:
[autofix.ci] apply automated fixes

maheshrajamani committed:
Revert changes to package-lock.json

maheshrajamani committed:
Revert changes to package-lock.json

autofix-ci[bot] committed:
[autofix.ci] apply automated fixes

maheshrajamani committed:
Merge branch 'model-training' of https://github.com/maheshrajamani/langflow into model-training

maheshrajamani committed:
Merge branch 'main' into model-training

maheshrajamani committed:
Fixed as per code style suggestion

autofix-ci[bot] committed:
[autofix.ci] apply automated fixes

maheshrajamani committed:
Fixed the code style to handle continue in else block

autofix-ci[bot] committed:
[autofix.ci] apply automated fixes

maheshrajamani committed:
Fixing code style issues

autofix-ci[bot] committed:
[autofix.ci] apply automated fixes

maheshrajamani committed:
Fix for code style issues

jordanrfrazier commented on src/backend/base/langflow/components/nvidia/nvidia_customizer.py:
```diff
@@ -0,0 +1,496 @@
+
+    def update_build_config(self, build_config, field_value, field_name=None):
+        """Updates the component's configuration based on the selected option or refresh button."""
+        models_url = f"{self.base_url}/v1/customization/configs"
```
When placing the component initially in a flow, it raises: 
```
Error while updating the Component

    Error refreshing model names: Request URL is missing an 'http://' or 'https://' protocol.
```

Probably just need error handling here for the empty case

jordanrfrazier commented on src/backend/base/langflow/components/nvidia/nvidia_customizer.py:
```diff
@@ -0,0 +1,496 @@
+            name="dataset",
+            display_name="Dataset",
+            info="Enter the dataset ID or name used to train the model",
+            value="test-data",
```
what is `test-data`?

jordanrfrazier replied:
I may even suggest leaving it empty

jordanrfrazier commented on src/backend/base/langflow/components/nvidia/nvidia_customizer.py:
```diff
@@ -0,0 +1,496 @@
+                    "finetuning_type": self.fine_tuning_type,
+                    "epochs": int(self.epochs),
+                    "batch_size": int(self.batch_size),
+                    "learning_rate": float(self.learning_rate),
```
```
  "hyperparameters": {
    "finetuning_type": "p_tuning",
    "training_type": "sft",
    "batch_size": 8,
    "epochs": 50,
    "learning_rate": 0.0001,
    "log_every_n_steps": 0,
    "val_check_interval": 0.01,
    "weight_decay": 0,
    "sft": {
      "hidden_dropout": 1,
      "attention_dropout": 1,
      "ffn_dropout": 1
    },
    "p_tuning": {
      "virtual_tokens": 0
    },
    "lora": {
      "adapter_dim": 8,
      "adapter_dropout": 1
    }
  },
  ```

Are we just supporting lora? What about p-tuning / sft? There's also a few other parameters missing -- are these intentionally left out?

jordanrfrazier commented on src/backend/base/langflow/components/nvidia/nvidia_customizer.py:
```diff
@@ -0,0 +1,496 @@
+
+        return build_config
+
+    async def customize(self) -> dict:
```
Needs more error handling around input parameters -- what if `base_url` is empty, misconfigured, etc. Are the nvidia api exceptions legible for a user?

maheshrajamani replied:
Handled.

jordanrfrazier commented on src/backend/base/langflow/components/nvidia/nvidia_customizer.py:
```diff
@@ -0,0 +1,496 @@
+        tenant = self.tenant_id if self.tenant_id else "tenant"
+        short_model_name = self.model_name.split("/")[-1]
+        customizations_url = f"{self.base_url}/v1/customization/jobs"
+        max_attempt_allowed = 10
```
Why 10? Why do we need retries at this level?

jordanrfrazier commented on src/backend/base/langflow/components/nvidia/nvidia_customizer.py:
```diff
@@ -0,0 +1,496 @@
+            await self.process_and_upload_dataset()
+
+        tenant = self.tenant_id if self.tenant_id else "tenant"
+        short_model_name = self.model_name.split("/")[-1]
```
This is splitting a user-supplied string; we should have a check on this as well

jordanrfrazier commented on src/backend/base/langflow/components/nvidia/nvidia_customizer.py:
```diff
@@ -0,0 +1,496 @@
+            # Build the data payload
+            data = {
+                "config": self.model_name,
+                "dataset": {"name": dataset_name, "namespace": tenant},
```
Hm, is `dataset` a dict? I see here it's a string - https://developer.nvidia.com/docs/nemo-microservices/customizer/source/api-reference.html

jordanrfrazier commented on src/backend/base/langflow/components/nvidia/nvidia_customizer.py:
```diff
@@ -0,0 +1,496 @@
+
+            # Build the data payload
+            data = {
+                "config": self.model_name,
```
should this be `name`?

jordanrfrazier replied:
I misunderstood -- I thought the it should be `"name": self.model_name`, not `"config"`. But I see the docs for this now.

jordanrfrazier commented on src/backend/base/langflow/components/nvidia/nvidia_customizer.py:
```diff
@@ -0,0 +1,496 @@
+            data = {
+                "config": self.model_name,
+                "dataset": {"name": dataset_name, "namespace": tenant},
+                "description": self.description,
```
This is just the langflow-hardcoded component description. I assume we either want a user-defined input here or leave it empty.

jordanrfrazier commented on src/backend/base/langflow/components/nvidia/nvidia_customizer.py:
```diff
@@ -0,0 +1,496 @@
+                    response = await client.post(customizations_url, headers=self.headers, json=data)
+
+                # If a conflict is returned, try again with an incremented attempt counter
+                if response.status_code == error_code_already_present:
```
Why do we handle 409s specially? I see where a user must create a dataset themselves where they define the dataset name. Why are we arbitrarily appending an attempt number to it?

jordanrfrazier replied:
(if this is some standard practice mentioned in the docs, I'd atleast recommend a randomly generated alphanumeric, not a counter)

jordanrfrazier commented on src/backend/base/langflow/components/nvidia/nvidia_customizer.py:
```diff
@@ -0,0 +1,496 @@
+
+                result_dict = {**result}
+                id_value = result_dict["id"]
+                result_dict["url"] = f"{customizations_url}/{id_value}/status"
```
Why are you adding this to the response?

jordanrfrazier commented on src/backend/base/langflow/components/nvidia/nvidia_customizer.py:
```diff
@@ -0,0 +1,496 @@
+                self.log(error_msg)
+                raise ValueError(error_msg) from exc
+            else:
+                return result_dict
```
Why is the `return` down here instead of just right after the successful response?

maheshrajamani replied:
I had to do this because on Ruff code styling issue ` TRY300 Consider moving this statement to an else block`

jordanrfrazier replied:
Stupid ruff

jordanrfrazier commented on src/backend/base/langflow/components/nvidia/nvidia_customizer.py:
```diff
@@ -0,0 +1,496 @@
+            except httpx.HTTPStatusError as exc:
+                # Check if the error is due to a 409 Conflict
+
+                if exc.response.status_code == error_code_already_present:
```
same question here

maheshrajamani replied:
Answered in the previous comment.

jordanrfrazier commented on src/backend/base/langflow/components/nvidia/nvidia_customizer.py:
```diff
@@ -0,0 +1,496 @@
+        try:
+            async with httpx.AsyncClient() as client:
+                response = await client.get(f"{url}/{namespace}")
+                self.log(f"returned data {response}")
```
remove or be more descriptive on what is returning data

jordanrfrazier commented on src/backend/base/langflow/components/nvidia/nvidia_customizer.py:
```diff
@@ -0,0 +1,496 @@
+            str: The dataset ID if found or created, or None if an error occurs.
+        """
+        dataset_name = self.get_dataset_name(user_dataset_name)
+        namespace = tenant_id if tenant_id else "tenant"
```
default value already sets this as `tenant` in the input param

jordanrfrazier replied:
can you link me to the concept of a `tenant` in the customizer docs? I couldn't find anything. Seems like this should just stay `namespace`

maheshrajamani replied:
Customizer has the concept of namespace. Our thought process is to use this for multi tenancy by using tenant id as namespace. Thinking loud believe this needs to be namespace since its NVIDIA component. @jeffreyscarpenter

jordanrfrazier commented on src/backend/base/langflow/components/nvidia/nvidia_customizer.py:
```diff
@@ -0,0 +1,496 @@
+                    create_response = await client.post(url, json=create_payload)
+                    create_response.raise_for_status()
+
+                return f"{namespace}/{dataset_name}"
```
what if it's some other non-404 error? Is this `return` supposed to be assuming that if it's not `404` then it exists and you just return the pre-determined namespace/dataset (repo-id) string?

maheshrajamani replied:
Added `response.raise_for_status()` to throw non 404 errors

jordanrfrazier commented on src/backend/base/langflow/components/nvidia/nvidia_customizer.py:
```diff
@@ -0,0 +1,496 @@
+        try:
+            # Inputs and repo setup
+            user_dataset_name = getattr(self, "dataset", None)
+            hf_api = HfApi(endpoint=f"{self.datastore_base_url}/v1/hf", token="")
```
Token should always be empty?

maheshrajamani replied:
The NeMo datastore micro service by default uses token to be passed as empty.

jordanrfrazier commented on src/backend/base/langflow/components/nvidia/nvidia_customizer.py:
```diff
@@ -0,0 +1,496 @@
+            raise ValueError(error_msg) from exc
+
+        try:
+            chunk_size = 100000  # Ensure chunk_size is an integer
```
why this number?

maheshrajamani replied:
When large dataset is to be uploaded, the data is chunked and stored as multiple files in the training folder.

jordanrfrazier replied:
Does it matter based on the microservice spec's? Does this need this be configurable for the MVP or can we leave it as is?

maheshrajamani replied:
NeMo micro services doesn't mandate anything for the size of these files. It helps with clearing the files (created in memory) once it's flushed to the data store.

jordanrfrazier commented on src/backend/base/langflow/components/nvidia/nvidia_customizer.py:
```diff
@@ -0,0 +1,496 @@
+            self.log(f"repo_id : {repo_id}")
+            if not repo_id:
+                err_msg = "repo_id must be provided."
+                raise ValueError(err_msg)
```
would suggest a different error phrasing. If the repo_id is empty, it's an internal error, not something the user can provide.

maheshrajamani replied:
Removed the error, the error handled in `get_repo_id` method

jordanrfrazier commented on src/backend/base/langflow/components/nvidia/nvidia_customizer.py:
```diff
@@ -0,0 +1,496 @@
+        # Generate a default dataset name using the current date and time
+        default_name = f"dataset-{datetime.now(timezone.utc).strftime('%Y%m%d%H%M%S')}"
+        # Use the user-provided name if available, otherwise the default
+        return user_dataset_name if user_dataset_name else default_name
```
It seems unintuitive that we'd generate a dataset name for the user. Why did you decide to do this?

maheshrajamani replied:
Will kill this logic and force user to provide dataset name.

jordanrfrazier commented on src/backend/base/langflow/components/nvidia/nvidia_customizer.py:
```diff
@@ -0,0 +1,496 @@
+            self.log(error_msg)
+            raise ValueError(error_msg) from e
+
+    async def process_and_upload_dataset(self) -> str:
```
I'd probably split up upload and process if possible

maheshrajamani replied:
upload file is maintained as separate method. Will change this method name as `process_dataset`

jordanrfrazier commented on src/backend/base/langflow/components/nvidia/nvidia_customizer.py:
```diff
@@ -0,0 +1,496 @@
+                raise ValueError(err_msg)
+
+            tasks = []
+            file_name_appender = user_dataset_name if user_dataset_name else "dataset"
```
how does this work with the whole auto-generated dataset name method?

maheshrajamani replied:
Yup redundant, removed it.

jordanrfrazier commented on src/backend/base/langflow/components/nvidia/nvidia_customizer.py:
```diff
@@ -0,0 +1,496 @@
+            name="training_data",
+            display_name="Training Data",
+            is_list=True,
+        ),
```
I see the example in the documentation assumes a user has data split into three directories: 
```
 To use the files during customization, you must add the appropriate path information to the files in training/, testing/, or validation/ directories.
```
Then it goes on to upload each individually like:
```
huggingface-cli upload --repo-type dataset default/sample-basic-test ~/tmp/sample_test_data/validation validation
```

In this implementation, it looks like you're splitting a single training data file into 10/90? I don't how this matches the documentation provided.

jordanrfrazier replied:
This is a complex component. I almost want to split it up into an ingest and process. 

That said, Langflow already has a problem with uploading large files. It helps a bit if we allow the user to pre-split their `validation` and `training` data files, as is done in the NeMo documentation. I would consider:
* Replacing  `training_data` input with a folder path that points to the base path `path/to/datasets`, and we'll take all files from `path/to/datasets/training/*` and `path/to/datasets/validation/*` for the respective customizer datas.
* Or, offer both training and validation inputs

maheshrajamani replied:
> * Replacing  `training_data` input with a folder path that points to the base path `path/to/datasets`, and we'll take all files from `path/to/datasets/training/*` and `path/to/datasets/validation/*` for the respective customizer datas.

The reason to not accept as files is because the component can accept data from different kind of sources like files, AstraDB, Arize dataset, or anything that emits `Data` object. 

> It helps a bit if we allow the user to pre-split their `validation` and `training` data files, as is done in the NeMo documentation

This was done purposefully so the user should not worry about the internals of NeMo micro services like data store, entity store, etc.

@jeffreyscarpenter Can you opine on this?

jeffreyscarpenter replied:
On the point about files, I agree that we need to consider adding a component that handles that part. The sample flow we have been using to demonstrate this contains custom components that do the file loading and generation of List[Data] with the required format i.e. "prompt", "completion". 

We should also include the flow we have been demonstrating as a starter project in this PR.

jordanrfrazier commented on src/backend/base/langflow/components/nvidia/nvidia_customizer.py:
```diff
@@ -0,0 +1,496 @@
+
+            # =====================================================
+            # STEP 4: Upload validation data (without chunking)
+            # =====================================================
```
What if this chunk is still large? Why don't we also chunk here based on some number of records

maheshrajamani replied:
Customizer only supports one file for validation. Have a bug reported with Nvidia,

jordanrfrazier commented on src/backend/base/langflow/components/nvidia/nvidia_customizer.py:
```diff
@@ -0,0 +1,496 @@
+
+            success_status_code = 200
+            if response.status_code == success_status_code:
+                logger.info("Chunk %s uploaded successfully!", self.chunk_number)
```
this log seems to indicate it's making a post for each chunk. this is only called once, iiuc, and the response is for registering the dataset, not uploading a chunk

maheshrajamani replied:
Fixed the error message. Missed with code refactoring

simonfraserduncan commented:
Some UX feedback:

### Customizer
- Update title to be NeMo Customizer (I believe @jeffreyscarpenter has already fixed or matched specs)
- Update description to be more useful (I believe @jeffreyscarpenter has already fixed)
- The base URL experience was confusing. If it's needed and not pre-populated, it should show on the default component UX. Ideally we reduce the number of required endpoints down
- The dataset field is confusing as is. I think we're trying to abstract away something without fully abstracting it. I would expect this to be more of a select experience similar to Astra DB collection. I'd ask: do we need to abstract this for our audience and those using a fine-tuning component? I think more transparency would be more intuitive.
-- If we are more explicit about the dataset, I'd expect to choose from existing datasets, or create a new one. This would not be a free text field, but a dropdown select. 
- I would try to avoid refreshes for the model name, training type, and fine tuning type. Follow the latest pattern for Astra DB, where refreshes happen automatically, and we show a manual refresh option within the list. 
- I would defer to other SME feedback about the advanced fields. 
- I know the job result URL is difficult to discover. Is the URL the same regardless of the job? We could surface it in the description or elsewhere through some form of 'Track job progress' action 
- Selection options that only have 1 item in a list are better off pre-selecting that option, but this is an existing issue elsewhere in Langflow. 
- Fine-tuning should be hyphenated 
- Eventually, we should update the fine-tuned model to show the dataset it was finetuned on, but doesn't feel essential for now. 

### Evaluator 
- Same issues with the base URLs
- LLM name placeholder value is confusing/unclear. Should it just say select a model? 
- Are tags meant to be single entry, or multiple entries?
- Evaluation type shouldn't have anything selected. It looks broken in that selecting the preselected item again shows new fields.

maheshrajamani committed:
Fixed as per review comments

maheshrajamani commented on src/backend/base/langflow/components/nvidia/nvidia_customizer.py:
```diff
@@ -0,0 +1,496 @@
+
+
+class NvidiaCustomizerComponent(Component):
+    display_name = "NVIDIA Customizer"
```
Named it similar to what was used for nvidia ingest component.

jeffreyscarpenter replied:
Please make it NeMo Customizer. this is what was agreed with NVIDIA product marketing folks

autofix-ci[bot] committed:
[autofix.ci] apply automated fixes

maheshrajamani committed:
Fixed as per review comments

maheshrajamani committed:
Merge branch 'model-training' of https://github.com/maheshrajamani/langflow into model-training

maheshrajamani committed:
Code improvement to evaluator based on reviews from yesterday.

autofix-ci[bot] committed:
[autofix.ci] apply automated fixes

maheshrajamani committed:
Code improvement to customizer:
1) Remove dataset name in the input
2) Add input field to provide the output model name

maheshrajamani committed:
Merge branch 'model-training' of https://github.com/maheshrajamani/langflow into model-training

autofix-ci[bot] committed:
[autofix.ci] apply automated fixes

maheshrajamani committed:
Refactored eval code for similar code changes done to evaluator.

maheshrajamani committed:
Merge branch 'model-training' of https://github.com/maheshrajamani/langflow into model-training

autofix-ci[bot] committed:
[autofix.ci] apply automated fixes

maheshrajamani committed:
Reordered the import for "I001 Import block is un-sorted or un-formatted"

maheshrajamani committed:
Merge branch 'model-training' of https://github.com/maheshrajamani/langflow into model-training

maheshrajamani committed:
Change the get_repo_id to create_namespace

autofix-ci[bot] committed:
[autofix.ci] apply automated fixes

maheshrajamani committed:
Fixed the rtpo

maheshrajamani committed:
Merge branch 'model-training' of https://github.com/maheshrajamani/langflow into model-training

jordanrfrazier commented on src/backend/base/langflow/components/nvidia/nvidia_customizer.py:
```diff
@@ -0,0 +1,480 @@
+        fine_tuned_model_name = self.fine_tuned_model_name
+
+        if not fine_tuned_model_name:
+            error_msg = "Provide fine tuned output model name to process"
```
I'd remove the `to process` from all of these messages. 

or, flip it to the pattern of `Missing x` rather than `Please provide x`

jordanrfrazier replied:
And all of these params should be required

jordanrfrazier commented on src/backend/base/langflow/components/nvidia/nvidia_customizer.py:
```diff
@@ -0,0 +1,480 @@
+            name="learning_rate",
+            display_name="Learning rate",
+            info="The number of samples used in each training iteration",
+            value=0.0001,
```
should batch size and learning rate be advanced?

jordanrfrazier commented on src/backend/base/langflow/components/nvidia/nvidia_customizer.py:
```diff
@@ -0,0 +1,480 @@
+        """Checks and creates namespace in datastore.
+
+        Args:
+            tenant_id (str): The tenant ID.
```
namespace vs. tenant naming

jordanrfrazier commented on src/backend/base/langflow/components/nvidia/nvidia_customizer.py:
```diff
@@ -0,0 +1,480 @@
+                    )
+                    tasks.append(task)
+                    chunk = []  # Reset the chunk
+                    self.chunk_number += 1
```
would be great to split these steps up and test the more interesting parts of the code, like the chunking and iteration

jordanrfrazier commented on src/backend/base/langflow/components/nvidia/nvidia_customizer.py:
```diff
@@ -0,0 +1,480 @@
+
+        return dataset_name
+
+    async def upload_chunk(self, chunk_df, chunk_number, file_name_prefix, repo_id, hf_api, is_validation):
```
We should parallelize the upload process in the future if we find it's taking a long time, especially since we re-ingest each time.

jordanrfrazier commented on src/backend/base/langflow/components/nvidia/nvidia_customizer.py:
```diff
@@ -0,0 +1,480 @@
+                is_validation = True
+                validation_df = pd.DataFrame(validation_records)
+                # Note: You will need to implement upload_validation to handle the
+                # upload of the full validation DataFrame.
```
what's this comment?

jordanrfrazier commented on src/backend/base/langflow/components/nvidia/nvidia_evaluator.py:
```diff
@@ -0,0 +1,707 @@
+
+        return repo_id
+
+    async def create_namespace(self, tenant_id: str):
```
Can share this code with the customizer

jordanrfrazier commented on src/backend/base/langflow/components/nvidia/nvidia_evaluator.py:
```diff
@@ -0,0 +1,707 @@
+        StrInput(
+            name="evaluator_base_url",
+            display_name="Evaluator Base URL",
+            info="The base URL of the NVIDIA NeMo Evaluator API.",
```
same comments on all these as above

jordanrfrazier commented on src/backend/base/langflow/components/nvidia/nvidia_evaluator.py:
```diff
@@ -0,0 +1,707 @@
+            advanced=True,
+        ),
+        StrInput(
+            name="tenant_id",
```
.

jordanrfrazier commented on src/backend/base/langflow/components/nvidia/nvidia_evaluator.py:
```diff
@@ -0,0 +1,707 @@
+            info="Select the model for evaluation",
+            options=[],  # Dynamically populated
+            refresh_button=True,
+        ),
```
We should create a template that uses the customizer and feeds the trained model into the evaluator

jordanrfrazier replied:
A template meaning an example for users to see how these two components can be used

jordanrfrazier commented on src/backend/base/langflow/components/nvidia/nvidia_evaluator.py:
```diff
@@ -0,0 +1,707 @@
+            info="The temperature to be used during generation sampling (0.0 to 2.0).",
+        ),
+        IntInput(
+            name="155_tokens_to_generate",
```
154?

jordanrfrazier commented on src/backend/base/langflow/components/nvidia/nvidia_customizer.py:
```diff
@@ -0,0 +1,480 @@
+        IntInput(
+            name="epochs",
+            display_name="Epochs",
+            info="Number of times to cycle through the training data. default : `5`",
```
And I've realized it's fine to leave out the "default: ..." lines in the `info` since they should be displayed in the value in the component, right?

jordanrfrazier commented on src/backend/base/langflow/components/nvidia/nvidia_evaluator.py:
```diff
@@ -0,0 +1,707 @@
+            options=["LM Evaluation Harness", "Similarity Metrics"],
+            value="LM Evaluation Harness",
+            real_time_refresh=True,  # Ensure dropdown triggers update on change
+        ),
```
This doesn't work as expected; the component initially appears without any of the additional fields for the LM evaluation harness. I'd suggest just leaving the initial value empty

jordanrfrazier commented on src/backend/base/langflow/components/nvidia/nvidia_evaluator.py:
```diff
@@ -0,0 +1,707 @@
+            display_name="Model to be evaluated",
+            info="Select the model for evaluation",
+            options=[],  # Dynamically populated
+            refresh_button=True,
```
This has a value of "Select a 000 llm name"

jordanrfrazier commented on src/backend/base/langflow/components/nvidia/nvidia_evaluator.py:
```diff
@@ -0,0 +1,707 @@
+        BoolInput(
+            name="150_greedy",
+            display_name="Few-shot Examples",
+            info="The number of few-shot examples before the input.",
```
"The number" vs. BoolInput

jordanrfrazier commented on src/backend/base/langflow/components/nvidia/nvidia_evaluator.py:
```diff
@@ -0,0 +1,707 @@
+            display_name="Tokens to Generate",
+            info="Max number of tokens to generate during inference.",
+            value=1024,
+        ),
```
Let's match other components as much as possible for this and other common params like temp and top k

```
        IntInput(
            name="max_tokens",
            display_name="Max Tokens",
            advanced=True,
            info="The maximum number of tokens to generate. Set to 0 for unlimited tokens.",
           value=1024,
        ),
```

jordanrfrazier commented on src/backend/base/langflow/components/nvidia/nvidia_evaluator.py:
```diff
@@ -0,0 +1,707 @@
+
+    def fetch_models(self):
+        """Fetch models from the specified API endpoint and return a list of model names."""
+        namespace = self.tenant_id if self.tenant_id else "tenant"
```
.

jordanrfrazier commented on src/backend/base/langflow/components/nvidia/nvidia_evaluator.py:
```diff
@@ -0,0 +1,707 @@
+            name="entity_service_base_url",
+            display_name="Entity service URL",
+            info="The base URL of the NVIDIA Entity service to obtain models that can be evaluated.",
+            advanced=True,
```
all required

jordanrfrazier commented on src/backend/base/langflow/components/nvidia/nvidia_evaluator.py:
```diff
@@ -0,0 +1,707 @@
+        DropdownInput(
+            name="310_run_inference",
+            display_name="Run Inference for response",
+            info="Select 'True' to run inference or 'False' to use an `response` field in the dataset.",
```
a `response`

maheshrajamani committed:
Changes based on feedback

autofix-ci[bot] committed:
[autofix.ci] apply automated fixes

maheshrajamani committed:
Used settings for nemo url's

maheshrajamani committed:
Merge branch 'model-training' of https://github.com/maheshrajamani/langflow into model-training

autofix-ci[bot] committed:
[autofix.ci] apply automated fixes

maheshrajamani committed:
Changed the component result as `Job Info`

maheshrajamani committed:
Merge branch 'model-training' of https://github.com/maheshrajamani/langflow into model-training

jordanrfrazier commented on src/backend/base/langflow/components/nvidia/nvidia_customizer.py:
```diff
@@ -0,0 +1,480 @@
+                            self.log(f"Updated training_type dropdown options: {training_types}")
+                            fine_tuning_type = selected_model.get("finetuning_types", [])
+                            build_config["fine_tuning_type"]["options"] = fine_tuning_type
+                            self.log(f"Updated fine_tuning_type dropdown options: {fine_tuning_type}")
```
We can cache as a map { model, [training, fine_tuning] }

jordanrfrazier commented on src/backend/base/langflow/components/nvidia/nvidia_customizer.py:
```diff
@@ -0,0 +1,480 @@
+            name="base_url",
+            display_name="Customizer Base URL",
+            info="The base URL of the NVIDIA Customizer API.",
+            advanced=True,
```
Okay, so `uuid` for the `dataset` name, meaning a `namespace`/`dataset` will be unique each run. The downside being that users have no way to easily inspect the files used for a specific output model, even if they're hosting the nemo data store (which, based on our default `localhost` urls for them, seems like that's we expect). That seems contradictory to me, but not a huge issue for demo purposes.

maheshrajamani committed:
Merge branch 'main' into model-training
